File: backend/app/core/money.py
"""
Money handling utilities using Decimal for precision
ESTÁNDAR NOUGRAM: Precisión financiera grado bancario
"""
from decimal import Decimal, ROUND_HALF_UP, ROUND_DOWN, ROUND_UP
from typing import Union, Optional
from app.core.logging import get_logger

logger = get_logger(__name__)

# Precisión estándar: 4 decimales para cálculos internos, 2 para display
INTERNAL_PRECISION = Decimal('0.0001')
DISPLAY_PRECISION = Decimal('0.01')

# COP no usa decimales en display pero sí en cálculos internos
COP_DISPLAY_PRECISION = Decimal('1')  # Sin decimales
COP_INTERNAL_PRECISION = Decimal('0.01')  # Precisión interna para cálculos

# ESTÁNDAR NOUGRAM: Usar ROUND_HALF_UP (no banker's rounding ROUND_HALF_EVEN)
# Esto asegura consistencia con Dinero.js que usa "round half up"
# Python usa ROUND_HALF_EVEN por defecto, debemos especificar explícitamente
ROUNDING_MODE = ROUND_HALF_UP  # NO ROUND_HALF_EVEN


class Money:
    """
    Clase inmutable para representar dinero con precisión Decimal
    ESTÁNDAR NOUGRAM: Todas las operaciones generan nuevas instancias
    """
    def __init__(self, amount: Union[float, int, str, Decimal], currency: str = "USD"):
        if isinstance(amount, Decimal):
            self._amount = amount
        elif isinstance(amount, str):
            self._amount = Decimal(amount)
        else:
            # Convertir float a string primero para evitar pérdida de precisión
            self._amount = Decimal(str(amount))
        
        self._currency = currency.upper()
        
        # Validar que el monto no sea negativo (a menos que sea explícitamente permitido)
        if self._amount < 0:
            logger.warning(f"Negative amount detected: {self._amount} {self._currency}")
    
    @property
    def amount(self) -> Decimal:
        """Retorna el monto como Decimal"""
        return self._amount
    
    @property
    def currency(self) -> str:
        """Retorna la moneda"""
        return self._currency
    
    def to_float(self) -> float:
        """Convierte a float (usar solo para compatibilidad con API)"""
        return float(self._amount)
    
    def to_int_cents(self) -> int:
        """Convierte a centavos/subunidades (para API)"""
        if self._currency == "COP":
            # COP ya está en unidades enteras
            return int(self._amount)
        # Otras monedas: multiplicar por 100
        return int(self._amount * 100)
    
    def quantize(self, precision: Optional[Decimal] = None) -> 'Money':
        """Redondea a la precisión especificada usando ROUND_HALF_UP"""
        if precision is None:
            precision = DISPLAY_PRECISION if self._currency != "COP" else COP_DISPLAY_PRECISION
        
        quantized = self._amount.quantize(precision, rounding=ROUNDING_MODE)
        return Money(quantized, self._currency)
    
    def add(self, other: 'Money') -> 'Money':
        """Suma dos montos (debe ser misma moneda)"""
        if self._currency != other._currency:
            raise ValueError(f"Cannot add {self._currency} and {other._currency}")
        return Money(self._amount + other._amount, self._currency)
    
    def subtract(self, other: 'Money') -> 'Money':
        """Resta dos montos (debe ser misma moneda)"""
        if self._currency != other._currency:
            raise ValueError(f"Cannot subtract {self._currency} and {other._currency}")
        return Money(self._amount - other._amount, self._currency)
    
    def multiply(self, multiplier: Union[float, Decimal]) -> 'Money':
        """Multiplica un monto por un escalar"""
        if isinstance(multiplier, float):
            multiplier = Decimal(str(multiplier))
        result = self._amount * multiplier
        return Money(result, self._currency)
    
    def divide(self, divisor: Union[float, Decimal]) -> 'Money':
        """Divide un monto por un escalar"""
        if isinstance(divisor, float):
            divisor = Decimal(str(divisor))
        if divisor == 0:
            raise ValueError("Cannot divide by zero")
        result = self._amount / divisor
        return Money(result, self._currency)
    
    def apply_percentage(self, percentage: Union[float, Decimal]) -> 'Money':
        """Aplica un porcentaje (ej: 19% de IVA)"""
        if isinstance(percentage, float):
            percentage = Decimal(str(percentage))
        multiplier = percentage / Decimal('100')
        return self.multiply(multiplier)
    
    def apply_margin(self, margin_percentage: Union[float, Decimal]) -> 'Money':
        """
        Calcula precio con margen: cost / (1 - margin)
        Ejemplo: cost = $100, margin = 40% → price = $100 / (1 - 0.40) = $166.67
        """
        if isinstance(margin_percentage, float):
            margin_percentage = Decimal(str(margin_percentage))
        
        if margin_percentage >= 100:
            raise ValueError("Margin cannot be >= 100%")
        
        margin_decimal = margin_percentage / Decimal('100')
        divisor = Decimal('1') - margin_decimal
        
        if divisor == 0:
            raise ValueError("Cannot apply 100% margin (division by zero)")
        
        result = self._amount / divisor
        return Money(result, self._currency).quantize()
    
    def __eq__(self, other):
        if not isinstance(other, Money):
            return False
        return self._amount == other._amount and self._currency == other._currency
    
    def __lt__(self, other):
        if not isinstance(other, Money):
            return NotImplemented
        if self._currency != other._currency:
            raise ValueError(f"Cannot compare {self._currency} and {other._currency}")
        return self._amount < other._amount
    
    def __le__(self, other):
        return self == other or self < other
    
    def __gt__(self, other):
        return not self <= other
    
    def __ge__(self, other):
        return not self < other
    
    def __repr__(self):
        return f"Money({self._amount}, '{self._currency}')"
    
    def __str__(self):
        return f"{self._amount} {self._currency}"


# Funciones helper para compatibilidad
def to_money(value: Union[float, int, str, Decimal], currency: str = "USD") -> Money:
    """Convierte un valor a Money"""
    return Money(value, currency)


def from_api(value: float, currency: str = "USD") -> Money:
    """Crea Money desde un valor de API (float)"""
    return Money(value, currency)


def to_api(money: Money) -> float:
    """Convierte Money a float para enviar a API"""
    return money.to_float()


def sum_money(amounts: list[Money]) -> Optional[Money]:
    """Suma una lista de Money (debe ser misma moneda)"""
    if not amounts:
        return None
    
    currency = amounts[0].currency
    total = Decimal('0')
    
    for amount in amounts:
        if amount.currency != currency:
            raise ValueError(f"All amounts must be in {currency}")
        total += amount.amount
    
    return Money(total, currency)

File: frontend/src/lib/money.ts
/**
 * Money handling utilities using dinero.js v2 for precision
 * ESTÁNDAR NOUGRAM: Sintaxis correcta de Dinero.js v2
 * 
 * IMPORTANTE: Dinero.js v2 usa:
 * - `dinero()` (minúscula) en lugar de `Dinero()`
 * - Importar monedas desde `@dinero.js/currencies`
 * - Tipo `Dinero<number>` en lugar de `Dinero.Dinero`
 */
import { dinero, add, multiply, type Dinero } from 'dinero.js';
import { USD, COP, EUR, ARS } from '@dinero.js/currencies';

// Configuración de monedas con objetos de Dinero.js v2
export const CURRENCY_CONFIG: Record<string, {
  currency: typeof USD;
  precision: number;
}> = {
  USD: { currency: USD, precision: USD.exponent },
  COP: { currency: COP, precision: COP.exponent },
  EUR: { currency: EUR, precision: EUR.exponent },
  ARS: { currency: ARS, precision: ARS.exponent },
};

/**
 * ESTÁNDAR NOUGRAM: Crea Dinero desde string Decimal del API
 * El backend serializa Decimal como string para mantener precisión
 * 
 * Ejemplo: "100.50" (USD) → dinero({ amount: 10050, currency: USD })
 * Ejemplo: "1000000" (COP) → dinero({ amount: 1000000, currency: COP })
 * 
 * Redondeo: Usa Math.round() que implementa "round half up" (no banker's rounding)
 * para consistencia con backend que usa ROUND_HALF_UP
 */
export function fromAPIString(
  apiString: string,
  currencyCode: string = 'USD'
): Dinero {
  const config = CURRENCY_CONFIG[currencyCode] || CURRENCY_CONFIG.USD;
  const factor = Math.pow(10, config.precision);
  
  // Parsear string a número con alta precisión
  const numericValue = parseFloat(apiString);
  
  if (isNaN(numericValue)) {
    throw new Error(`Invalid money string: ${apiString}`);
  }
  
  // Convertir a unidades menores (centavos/subunidades)
  // Math.round() usa "round half up" (no banker's rounding)
  const amount = Math.round(numericValue * factor);
  
  // Sintaxis Dinero.js v2: dinero({ amount, currency })
  // @ts-ignore - dinero.js acepta objetos de moneda aunque la declaración dice string
  return dinero({ amount, currency: config.currency });
}

/**
 * Crea Dinero desde número (compatibilidad hacia atrás)
 * @deprecated Usar fromAPIString cuando backend envíe strings
 */
export function fromAPI(
  amount: number,
  currencyCode: string = 'USD'
): Dinero {
  const config = CURRENCY_CONFIG[currencyCode] || CURRENCY_CONFIG.USD;
  const factor = Math.pow(10, config.precision);
  const roundedAmount = Math.round(amount * factor);
  
  // @ts-ignore - dinero.js acepta objetos de moneda aunque la declaración dice string
  return dinero({ amount: roundedAmount, currency: config.currency });
}

/**
 * Convierte Dinero a string para enviar al API
 * ESTÁNDAR NOUGRAM: Backend espera strings Decimal
 */
export function toAPIString(dinero: Dinero): string {
  // Dinero.js v2: usar toJSON() para obtener amount y currency
  const json = dinero.toJSON();
  const amount = json.amount;
  const currencyCode = json.currency.code;
  const config = CURRENCY_CONFIG[currencyCode] || CURRENCY_CONFIG.USD;
  const factor = Math.pow(10, config.precision);
  const value = amount / factor;
  
  return value.toFixed(config.precision);
}

/**
 * Obtiene el valor numérico de Dinero (para compatibilidad)
 */
export function toAPI(dinero: Dinero): number {
  // Dinero.js v2: usar toJSON() para obtener amount
  const json = dinero.toJSON();
  const amount = json.amount;
  const currencyCode = json.currency.code;
  const config = CURRENCY_CONFIG[currencyCode] || CURRENCY_CONFIG.USD;
  const factor = Math.pow(10, config.precision);
  
  return amount / factor;
}

/**
 * Suma múltiples Dinero (debe ser misma moneda)
 */
export function sumMoney(amounts: Dinero[]): Dinero | null {
  if (amounts.length === 0) return null;
  
  return amounts.reduce((total, amount) => {
    // Dinero.js v2: usar función add() importada directamente
    return add(total, amount);
  });
}

/**
 * Multiplica Dinero por un escalar
 */
export function multiplyMoney(dineroObj: Dinero, multiplier: number): Dinero {
  // Dinero.js v2: usar toJSON() para obtener amount, luego crear Dinero con multiplicador
  const json = dineroObj.toJSON();
  const newAmount = Math.round(json.amount * multiplier);
  const config = CURRENCY_CONFIG[json.currency.code] || CURRENCY_CONFIG.USD;
  // Crear nuevo Dinero con el amount multiplicado usando dinero()
  return dinero({ amount: newAmount, currency: config.currency });
}

/**
 * Divide Dinero por un escalar
 */
export function divideMoney(dineroObj: Dinero, divisor: number): Dinero {
  if (divisor === 0) {
    throw new Error('Cannot divide by zero');
  }
  // Dinero.js v2: usar toJSON() para obtener amount, luego crear Dinero con divisor
  const json = dineroObj.toJSON();
  const newAmount = Math.round(json.amount / divisor);
  const config = CURRENCY_CONFIG[json.currency.code] || CURRENCY_CONFIG.USD;
  // Crear nuevo Dinero con el amount dividido usando dinero()
  return dinero({ amount: newAmount, currency: config.currency });
}

/**
 * Aplica un porcentaje a Dinero (ej: 19% de IVA)
 */
export function applyPercentage(dineroObj: Dinero, percentage: number): Dinero {
  // Aplicar porcentaje: amount * (percentage / 100)
  const multiplier = percentage / 100;
  return multiplyMoney(dineroObj, multiplier);
}

/**
 * Calcula precio con margen: cost / (1 - margin)
 * ESTÁNDAR NOUGRAM: Usa redondeo "half up" para consistencia con backend
 */
export function applyMargin(
  cost: Dinero,
  marginPercentage: number
): Dinero {
  if (marginPercentage >= 100) {
    throw new Error('Margin cannot be >= 100%');
  }
  
  const marginDecimal = marginPercentage / 100;
  const divisor = 1 - marginDecimal;
  const multiplier = 1 / divisor;
  
  // Dinero.js v2: usar multiplyMoney helper
  return multiplyMoney(cost, multiplier);
}

/**
 * Formatea Dinero usando localización específica
 * ESTÁNDAR NOUGRAM: Usa locale 'es-CO' para formateo explícito
 */
export function formatCurrency(
  dinero: Dinero,
  locale: string = 'es-CO',
  options?: { showSymbol?: boolean }
): string {
  // Dinero.js v2: usar toJSON() para obtener currency y amount
  const json = dinero.toJSON();
  const currency = json.currency.code;
  const showSymbol = options?.showSymbol !== false;
  
  // Formato específico por moneda
  if (currency === 'COP') {
    // COP: sin decimales, punto como separador de miles
    const amount = json.amount; // Ya está en unidades enteras
    const formatted = amount.toLocaleString('es-CO', {
      minimumFractionDigits: 0,
      maximumFractionDigits: 0,
    });
    return showSymbol ? `$${formatted}` : formatted;
  } else {
    // Otras monedas: con decimales
    const currencyCode = json.currency.code;
    const config = CURRENCY_CONFIG[currencyCode] || CURRENCY_CONFIG.USD;
    const factor = Math.pow(10, config.precision);
    const value = json.amount / factor;
    if (showSymbol) {
      return value.toLocaleString(locale, {
        style: 'currency',
        currency: currency,
        minimumFractionDigits: 2,
        maximumFractionDigits: 2,
      });
    } else {
      return value.toLocaleString(locale, {
        style: 'decimal',
        minimumFractionDigits: 2,
        maximumFractionDigits: 2,
      });
    }
  }
}

/**
 * Helper para crear Dinero desde string (input de usuario)
 */
export function fromString(value: string, currencyCode: string = 'USD'): Dinero {
  // Remover símbolos y espacios
  const cleanValue = value.replace(/[^\d.,-]/g, '');
  
  // Detectar separador decimal
  const hasComma = cleanValue.includes(',');
  const hasDot = cleanValue.includes('.');
  
  let numericValue: number;
  
  if (hasComma && hasDot) {
    // Formato europeo: 1.234,56
    const parts = cleanValue.split(',');
    numericValue = parseFloat(parts[0].replace(/\./g, '') + '.' + parts[1]);
  } else if (hasComma) {
    // Solo coma: 1234,56
    numericValue = parseFloat(cleanValue.replace(',', '.'));
  } else {
    // Solo punto o sin separador: 1234.56 o 1234
    numericValue = parseFloat(cleanValue);
  }
  
  return fromAPI(numericValue, currencyCode);
}
